### PR TITLE
Update the audit sources support for JS, Python, PHP, Go & CSharp

### PR DESCRIPTION
**Pull request by dbMundada**
- For each language what support tobe  added mentioned in doc: https://privado-ai.atlassian.net/wiki/spaces/arc/pages/979009551/DED+Audit+Report+Data
- Add support for the Csharp, PHP, Ruby, Golang & Python
- Moved the common utilities in DataElementDiscoveryUtils
- Created single functions getIdentifiers, getFieldAccessIdentifier, getMemberUsingClassName to add support for those inline
- Added some common filters & language specific filters to remove noise from audit sources.
- Also, handled possibel error comes due to invalid regex pattern in **XX.typeFullName**

### CODE REVIEW

**Comment by karan-batavia** [comment]
Can you add a comment stating the enhancements done in the PR? It's a considerably large one, I might need some context for review. @dbMundada

File: src/main/scala/ai/privado/audit/AuditReportEntryPoint.scala [hunk 1]
```diff
@@ -85,10 +85,8 @@ object AuditReportEntryPoint {
     lang: Language = Language.JAVA
   ): Workbook = {
     lang match {
-      case Language.JAVASCRIPT =>
-        getAuditWorkbookJS(xtocpg, taggerCache, repoPath, auditCache, ruleCache)
-      case Language.GO | Language.PYTHON =>
-        getAuditWorkbookGoAndPy(xtocpg, taggerCache, repoPath, auditCache, ruleCache)
+      case Language.JAVASCRIPT | Language.PHP | Language.CSHARP | Language.PYTHON | Language.GO | Language.RUBY =>
+        getAuditWorkbookForLanguage(xtocpg, taggerCache, repoPath, auditCache, ruleCache, lang)
       case Language.JAVA | Language.KOTLIN =>
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/AuditReportEntryPoint.scala:90]: Can't we not use `getAuditWorkbookForLanguage` for Java and Kotlin as well?

File: src/main/scala/ai/privado/audit/AuditReportEntryPoint.scala [hunk 2]
```diff
@@ -107,7 +105,7 @@ object AuditReportEntryPoint {
   ): Workbook = {
     val workbook: Workbook = new XSSFWorkbook()
     // Set Element Discovery Data into Sheet
-    val dataElementDiscoveryData = DataElementDiscovery.processDataElementDiscovery(xtocpg, taggerCache)
+    val dataElementDiscoveryData = DataElementDiscoveryJava.processDataElementDiscovery(xtocpg, taggerCache)
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/AuditReportEntryPoint.scala:108]: I presume DataElementDiscovery is the same for every language with minor differences, why are we creating a `DataElementDiscoveryJava` object?

Comment (karan-batavia) [reply src/main/scala/ai/privado/audit/AuditReportEntryPoint.scala:108]: Suggestion would be to have a single class which should be instantiated.

File: src/main/scala/ai/privado/audit/DataElementDiscovery.scala [hunk 3]
```diff
@@ -41,9 +74,271 @@ object DataElementDiscoveryUtils {
...
+
+  // Add Identifiers in workbookResult
+  def getIdentifiers(
+    xtocpg: Try[Cpg],
+    workbookResult: ListBuffer[List[String]],
+    lang: Language
+  ): ListBuffer[List[String]] = {
+    val identifiers = xtocpg match {
+      case Success(cpg) => {
+        println(cpg.identifier.filter(i => i.name.length > 2).l.size)
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/DataElementDiscovery.scala:86]: Can you remove this println please?

File: src/main/scala/ai/privado/audit/DataElementDiscovery.scala [hunk 4]
```diff
@@ -41,9 +74,271 @@ object DataElementDiscoveryUtils {
...
+    lang: Language
+  ): ListBuffer[List[String]] = {
+    val fieldIdentifiers = xtocpg match {
+      case Success(cpg) => {
+        cpg.fieldAccess.fieldIdentifier
+          .filter(i => i.canonicalName.length > 2)
+          .filter(i => !filterCommonVarsStartsWithArr.exists(xx => i.canonicalName.startsWith(xx)))
+          .filter(i => !i.canonicalName.matches(filterCommonLangTypes))
+          .filter(i => !i.canonicalName.matches(filterCommonVars))
+          .filter(i => !i.canonicalName.matches(DataElementDiscoveryUtils.getLanguageSpecificFilters(lang)))
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/DataElementDiscovery.scala:154]: Can we add some comments in the code explaining these filters? Looks like there are many, else we could add methods with descriptive names.

File: src/main/scala/ai/privado/audit/DataElementDiscovery.scala [hunk 5]
```diff
@@ -41,9 +74,271 @@ object DataElementDiscoveryUtils {
...
+          .l
+      }
+      case Failure(ex) => {
+        ex.printStackTrace()
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/DataElementDiscovery.scala:158]: This would print to console, we don't want that.

Comment (karan-batavia) [reply src/main/scala/ai/privado/audit/DataElementDiscovery.scala:158]: Can you add a `logger.debug` instead?

File: src/main/scala/ai/privado/audit/DataElementDiscovery.scala [hunk 6]
```diff
@@ -41,9 +74,271 @@ object DataElementDiscoveryUtils {
...
+      case Success(cpg) => {
+        classNameSet.foreach(className => {
+          try {
+            cpg.typeDecl
+              .where(_.fullName(className))
+              .foreach(typeDeclNode => {
+                if (typeDeclNode.member.nonEmpty) {
+                  val members =
+                    typeDeclNode.member
+                      .filter(i => i.name.length > 2)
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/DataElementDiscovery.scala:219]: Same for these filters.

File: src/main/scala/ai/privado/audit/DataElementDiscovery.scala [hunk 7]
```diff
@@ -41,9 +74,271 @@ object DataElementDiscoveryUtils {
...
+                }
+              })
+          } catch {
+            case ex: Exception =>
+              logger.debug(f"Skipping the class ${className} due to invalid regex issue", ex)
+          }
+        })
+      }
+      case Failure(exception) => {
+        println("Failed to process member name from cpg")
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/DataElementDiscovery.scala:236]: Need to remove this `println`.

File: src/main/scala/ai/privado/audit/DataElementDiscovery.scala [hunk 8]
```diff
@@ -41,9 +74,271 @@ object DataElementDiscoveryUtils {
...
+        logger.debug("Failed to process member name from cpg", exception)
+        println(exception.printStackTrace())
```
Comment (karan-batavia) [inline src/main/scala/ai/privado/audit/DataElementDiscovery.scala:238]: Same for this.

**Review by karan-batavia — CHANGES_REQUESTED** [review]

File: src/main/scala/ai/privado/audit/AuditReportEntryPoint.scala (hunk 2 above)
Comment (dbMundada) [reply src/main/scala/ai/privado/audit/AuditReportEntryPoint.scala]: No for Java/Kotlin we written some specific logic to detect the files within cpg.typeDecl
Additionally it has Java/kotlin specific filtering logic implemented.
Also, some specific logic for derivedClass which is not applicable for other languages.

So, let it be separate for now.

**Review by karan-batavia — APPROVED** [review]
LGTM

**Pull request merged**